### PR TITLE
Nightwatch for Drupal

### PR DESCRIPTION
**Pull request by WengerK**
I test this code with a mounted volume into Watchdreamer.

_docker-compose.override.yml_
```
  # Drupal test server
  test:
    hostname: test
    ports:
      - "8888:8888"
    volumes:
      - /Users/wengerk/www/contrib/docker/docker-php-dev/scripts/docker-as-drupal:/usr/local/bin/docker-as-drupal
```

I also suggest to change the codeshipsteps.yml to

```
>-
  bash -c "docker-as-wait --mysql test:8888 browser:4444 &&
    drush cset wd_test.settings mode 'nightwatch' -y &&
    docker-as-drupal nightwatch --group=wd"
```

But I'm maybe wrong, can we do that:

```
command: docker-as-wait --mysql test:8888 browser:4444 -- docker-as-drupal nightwatch --group=wd --skip-styleguide-build

```

I don't understand how to tests in a realife codeship instance.

### CODE REVIEW

File: scripts/docker-as-drupal [hunk 1]
```diff
@@ -898,6 +910,117 @@ elif [ "$1" = "behat" ]; then
...
+  SKIP_DEPENDENCIES=0
+  SKIP_STYLEGUIDE=0
+  SKIP_DB_RESET=0
+  SKIP_DEFAULT_CONTENT=0
+  NIGHTWATCH_GROUPS=()
+  NIGHTWATCH_ARGV=()
+
+  while [ $# -gt 0 ]; do
+    case "$1" in
+      --skip-dependencies)
```
Comment (yannlugrin) [inline scripts/docker-as-drupal]: I prefer `--with-dependencies`, the case where we don't have already installed it is not comon

File: scripts/docker-as-drupal [hunk 2]
```diff
@@ -898,6 +910,117 @@ elif [ "$1" = "behat" ]; then
...
+  drush pmu $TEST_DISABLE_MODULES -y
+
+  # Enable modules that must be loaded for tests
+  drush en $TEST_ENABLE_MODULES -y
+
+  # Rebuild Drupal cache
+  drush cache-rebuild
+
+  # Build styleguide
+  if yarn_is_installed && [ $SKIP_STYLEGUIDE -eq 0 ]; then
```
Comment (yannlugrin) [inline scripts/docker-as-drupal]: We can remove `yarn_is_installed` because nightwatch require it anyway.
```suggestion
  if [ $SKIP_STYLEGUIDE -eq 0 ]; then
```

File: scripts/docker-as-drupal [hunk 3]
```diff
@@ -898,6 +910,117 @@ elif [ "$1" = "behat" ]; then
...
+
+  # Install Nightwatch in the core directory
+  yarn install
+
+  # Run tests
+  yarn test:nightwatch ${NIGHTWATCH_GROUPS[@]} ${NIGHTWATCH_ARGV[@]}
+  nightwatch_exit=$?
+
+  # move back in the root directory
+  cd ../../
```
Comment (yannlugrin) [inline scripts/docker-as-drupal]: AS note, we could use subshell here to avoid that. But this way is okay.

File: scripts/docker-as-drupal (hunk 1 above)
Comment (WengerK) [reply scripts/docker-as-drupal]: I use the same pattern you use for bootstrap

Comment (yannlugrin) [reply scripts/docker-as-drupal]: Locally, bootstrap command is used to install dependencies, on codeship they are installed on build. Most time you don't have to update them when you run tests (never have to do on codeship), so it's why in tests commands I pefer `--with-dependencies` but have `--skip-dependencies` in bootstrap command.

**Comment by WengerK** [comment]
We have a big issue :/

`drush cset wd_test.settings mode 'nightwatch' -y` can't be done in the `> command` of `codeshipsteps.yml` as we need the module `wd_test` to be enabled - this module is enabled in the `docker-as-drupal nightwatch` ...

File: scripts/docker-as-drupal (hunk 3 above)
Comment (yannlugrin) [reply scripts/docker-as-drupal]:
```
(
  cd web/core

  # ...
)
```

**Comment by yannlugrin** [comment]
Could we do that in the `nightwatch` yarn command ?

File: scripts/docker-as-drupal (hunk 1 above)
Comment (WengerK) [reply scripts/docker-as-drupal]: not the Nightwatch dependencies. They should be installed in web/core which are not installed in our boostrap.

**Comment by WengerK** [comment]
I would prefer not having to add a `package.json` script specific by project (which we will forget). what about a new 

```
      DRUPAL_TEST_CONFIG_SET: >-
        wd_test.settings mode 'nightwatch'
```

**Comment by WengerK** [comment]
I do the changes (it was painfull to deal with multiline ENV variables which have spaces) ....

usage

singleline:
```
      DRUPAL_NIGHTWATCH_CONFIG_SET: >-
        wd_test.settings mode 'nightwatch'
```

multiline:
```
      DRUPAL_NIGHTWATCH_CONFIG_SET: >-
        wd_test.settings mode 'nightwatch' ;
        wd_test.settings enable 1 ;

```

File: scripts/docker-as-drupal [hunk 4]
```diff
@@ -898,6 +911,125 @@ elif [ "$1" = "behat" ]; then
...
+  fi
+
+  # Disable modules that must not be loaded for tests
+  drush pmu $TEST_DISABLE_MODULES -y
+
+  # Enable modules that must be loaded for tests
+  drush en $TEST_ENABLE_MODULES -y
+
+  if [ ! -z "$DRUPAL_NIGHTWATCH_CONFIG_SET" ]; then
+    IFS=';' read -r -a configs <<< "$DRUPAL_NIGHTWATCH_CONFIG_SET"
```
Comment (yannlugrin) [inline scripts/docker-as-drupal]: Why don't use default "new line" as list separator?

Comment (WengerK) [reply scripts/docker-as-drupal]: because IFS don't support more than 1 char (using `\n` will split on `\` & `n`)

Comment (yannlugrin) [reply scripts/docker-as-drupal]: I don't change it for `$REQUIRED_DIRECTORIES`, just loop on the variable without need to use `read` or change `IFS` value.

File: Dockerfile [hunk 5]
```diff
@@ -67,7 +67,7 @@ RUN set -ex; \
     imagemagick \
     jq \
     libpng-dev \
-    mysql-client \
+    mariadb-client \
```
Comment (yannlugrin) [inline Dockerfile]: This is not related to nightwatch, just a change in base image, mysql client is replaced by mariadb client.

File: scripts/docker-as-drupal (hunk 3 above)
Comment (gido) [reply scripts/docker-as-drupal]: I didn't know either :)

File: Dockerfile (hunk 5 above)
Comment (gido) [reply Dockerfile]: can we know why?

Comment (yannlugrin) [reply Dockerfile]: I don't know why, base distribution probably change that regarding the license. It's for the same reason that RMF website deploy failed this week.

File: README.md [hunk 6]
```diff
@@ -390,6 +390,22 @@ docker-compose exec test docker-as-drupal behat [options]
   -- <...>                 # any behat valid args
 ```
 
+*nightwatch* setup database and settings properly then run nightwatch command. `--group` is required
+as Drupal nitghtwatch tests doesn't pass without more config.
```
Comment (gido) [inline README.md]:
```suggestion
as Drupal nightwatch tests doesn't pass without more config.
```
